build-support: Fix error when building images with many layers

koenw opened the pull request:
When building a docker image using `dockertools.buildLayeredImage`, the resulting image layers are passed to `jq` through the command line. When building an image with too many layers this would exceed the maximum command line argument length.

Hence, we store the list of layers in the Nix store and pass them to `jq` as a file argument using `--slurpfile`.

Fixes #140908.

###### Motivation for this change
To fix  #140908.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Build and loaded a docker image that failed previously
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

roberth reviewed (commented):
Just some superficial things left.
Could you also squash the commit and prefix with `dockerTools:` instead? It's more recognizable.

hlolli commented:
I bumped into this jq Argument list issue, this PR fixed it! Do you want help getting this PR across the finishing line @koenw ?

koenw committed:
build-support: Fix error when building images with many layers

When building a docker image using `dockertools.buildLayeredImage`, the
resulting image layers are passed to `jq` through the command line. When
building an image with too many layers this would exceed the maximum
command line argument length.

Hence, we store the list of layers in the Nix store and pass them to
`jq` as a file argument using `--slurpfile`.

Fixes #140908.

koenw commented on pkgs/build-support/docker/default.nix:
```diff
@@ -893,11 +893,6 @@ rec {
-          if [[ "$created" != "now" ]]; then
-              created="$(date -Iseconds -d "$created")"
-          fi
-
```
I don't remember... I moved it back and squashed the commits.

koenw commented on pkgs/build-support/docker/default.nix:
```diff
@@ -883,7 +883,7 @@ rec {
                 lib.head (lib.strings.splitString "-" (baseNameOf conf.outPath));
             paths = buildPackages.referencesByPopularity overallClosure;
             nativeBuildInputs = [ jq ];
-          } ''
```
Done.

koenw commented on pkgs/build-support/docker/default.nix:
```diff
@@ -935,33 +930,39 @@ rec {
-                --argjson maxLayers "$availableLayers"
-          )"
+          paths |
+            jq -sR '
+              rtrimstr("\n") | split("\n")
+                | (.[:$maxLayers-1] | map([.])) + [ .[$maxLayers-1:] ]
+                | map(select(length > 0))
+            ' \
+              --argjson maxLayers "$availableLayers" > store_layers.json
```
I'm sorry, it's unclear to me from the way GitHub displays this diff what exactly it is you're suggesting (except perhaps changing a random space into a \t but I doubt it). However it seems it can be directly applied by GitHub so feel free.

roberth replied:
It's just the two spaces I think.

> can be directly applied by GitHub

Yet, it doesn't give me the button to apply it. Let me try another (same) suggestion:

```suggestion
            jq -sR '
              rtrimstr("\n") | split("\n")
                | (.[:$maxLayers-1] | map([.])) + [ .[$maxLayers-1:] ]
                | map(select(length > 0))
              ' \
              --argjson maxLayers "$availableLayers" > store_layers.json
```

koenw replied:
oooooh, now I see! I was distracted by all the red "highlights" that weren't changes...

Because you wanted to squash the commits I've updated the branch manually rather than through GitHub's suggestion.

koenw commented:
> I bumped into this jq Argument list issue, this PR fixed it! Do you want help getting this PR across the finishing line @koenw ?

Thanks, I forgot about this PR :)

@roberth I implemented most of your suggestions (except one - see notes). Can you do a re-review please? :)

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -935,33 +935,34 @@ rec {
+              rtrimstr("\n") | split("\n")
+                | (.[:$maxLayers-1] | map([.])) + [ .[$maxLayers-1:] ]
+                | map(select(length > 0))
+            ' \
```
Maybe it has to be a new comment on a line that's from the latest state?

```suggestion
              ' \
```

roberth replied:
Nope. Maybe you haven't given permission for maintainers to push to your branch?

roberth commented:
@ofborg test docker-tools

roberth commented:
Thank you @koenw!
